beacon-chain/types/block.go

```diff
@@ -128,3 +140,103 @@ func (b *Block) Attestations() []*pb.AttestationRecord {
 func (b *Block) Timestamp() (time.Time, error) {
 	return ptypes.Timestamp(b.data.Timestamp)
 }
+
+// isSlotValid compares the slot to the system clock to determine if the block is valid
+func (b *Block) isSlotValid() bool {
+	slotDuration := time.Duration(b.SlotNumber()*params.SlotDuration) * time.Second
+	validTimeThreshold := genesisTime.Add(slotDuration)
+
+	return clock.Now().After(validTimeThreshold)
+}
+
+// IsValid is called to decide if an incoming p2p block can be processed into the chain's block trie,
+// it checks time stamp, beacon chain parent block hash. It also checks pow chain reference hash if it's a validator.
+func (b *Block) IsValid(aState *ActiveState, cState *CrystallizedState) bool {
+	_, err := b.Hash()
+	if err != nil {
+		log.Debugf("Could not hash incoming block: %v", err)
+		return false
+	}
+
+	if b.SlotNumber() == 0 {
+		log.Debugf("Can not process a genesis block")
+		return false
+	}
+
+	if !b.isSlotValid() {
+		log.Debugf("slot of block is too high: %d", b.SlotNumber())
+		return false
+	}
+
+	for index := range b.Attestations() {
+		if !b.isAttestationValid(index, aState, cState) {
+			return false
```

Comment:
Should we also `log.Debugf` here to show which attestation is invalid?

Reply:
yeah, good idea